### PR TITLE
Ensure NestedModules property gets populated by Test-ModuleManifest

### LINKED ISSUE #7833: Test-ModuleManifest does not populate the NestedModules field
When I run the Test-ModuleManifest cmdlet with a psd1 file that has a NestedModules field, it returns null for NestedModules (even though the dll exists and is in the correct location).  This results in Update-ModuleManifest commenting out the existing NestedModule fields when I try to update just the version.  This is causing massive issues with our versioning scripts when trying to update from PowerShell 5.1 to PowerShell 6.0.  Running this in PowerShell 5.1 with the same files works as expected.  When I pulled down PowerShell to try to debug this, I saw that NestedModules was set correctly here: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs#L2041, but not set in moduleInfo before moduleInfo was returned here: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs#L2800.

Steps to reproduce
------------------

```powershell
$a = Test-ModuleManifest -Path <path to module with filled out nestedmodules field>
$a.NestedModule # Returns nothing
```

Expected behavior
-----------------

```none
$a = Test-ModuleManifest -Path <path to module with filled out nestedmodules field>
$a.NestedModule # Returns correct Nested Modules
```

Actual behavior
---------------
Returns nothing
```none

```

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.1.0-preview.4
PSEdition                      Core
GitCommitId                    6.1.0-preview.4
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
## PR Summary

This is a regression introduced by https://github.com/PowerShell/PowerShell/pull/7145 as an optimization where no further analysis of the module is done if none of the exported capabilities have a wildcard.  However, the NestedModules property hasn't been populated yet, so it remains empty.  This is a regression as previously this property is always populated if the manifest contains a value.

Fix is to create PSModuleInfo entries for each specified NestedModule even though we don't do actual loading and analysis.  Creating the fake NestedModule is similar to how the code already creates fake RequiredModules when not importing.

Also some minor cleanup of repetitive code in the tests.

Fix https://github.com/PowerShell/PowerShell/issues/7833

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by rjmholt** [comment]
This seems like the right way to efficiently build/return the nested module info objects. But for the sake of having considered other ideas, wouldn't users expect a fully live nested module info object here?

Since the fake module info is the same type that would normally be returned by `Test-ModuleManifest` but is hollow, the fact that we took an optimised path means you now don't get back fields like `ModuleBase` and have no idea why. Unlike `RequiredModules`, `NestedModules` really forms a tree, and you might want to traverse that -- but now you wouldn't be able to consistently expect to be able to use that behaviour.

[Further down in the same method](https://github.com/PowerShell/PowerShell/blob/76c9da189dd5990010fbce3d44890d9659be9b0e/src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs#L2927-L2949) when nested modules are read, we actually go and read their manifest and load a module info.

Creating fake info objects from the nested module specifications is certainly more efficient, but I think it's a breaking change from previous (pre-regression) behaviour. See Windows PowerShell's version of the same:
```
PS C:\Users\roholt\Documents\Dev\sandbox\mod> $x = Test-ModuleManifest .\mod.psd1
PS C:\Users\roholt\Documents\Dev\sandbox\mod> $x.NestedModules |fl *


LogPipelineExecutionDetails : False
Name                        : x
Path                        : C:\Users\roholt\Documents\Dev\sandbox\mod\x.psm1
ImplementingAssembly        :
Definition                  :
Description                 :
Guid                        : 00000000-0000-0000-0000-000000000000
HelpInfoUri                 :
ModuleBase                  : C:\Users\roholt\Documents\Dev\sandbox\mod
PrivateData                 :
Tags                        : {}
ProjectUri                  :
IconUri                     :
LicenseUri                  :
ReleaseNotes                :
RepositorySourceLocation    :
Version                     : 0.0
ModuleType                  : Script
Author                      :
AccessMode                  : ReadWrite
ClrVersion                  :
CompanyName                 :
Copyright                   :
DotNetFrameworkVersion      :
ExportedFunctions           : {[Banana, Banana]}
Prefix                      :
ExportedCmdlets             : {}
ExportedCommands            : {[Banana, Banana]}
FileList                    : {}
CompatiblePSEditions        : {}
ModuleList                  : {}
NestedModules               : {}
PowerShellHostName          :
PowerShellHostVersion       :
PowerShellVersion           :
ProcessorArchitecture       : None
Scripts                     : {}
RequiredAssemblies          : {}
RequiredModules             : {}
RootModule                  :
ExportedVariables           : {}
ExportedAliases             : {}
ExportedWorkflows           : {}
ExportedDscResources        : {}
SessionState                :
OnRemove                    :
ExportedFormatFiles         : {}
ExportedTypeFiles           : {}
```

I think we may be committed to doing the inefficient thing.

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 1]
```diff
@@ -2797,6 +2797,20 @@ internal bool LoadModuleManifestData(ExternalScriptInfo scriptInfo, ManifestProc
 
                 if (!needToAnalyzeScriptModules)
                 {
+                    // need to add nested modules to the manifestInfo when no more analysis needs to be done
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: // Add nested modules ...

**Review by PaulHigin — COMMENTED** [review]
LGTM

**Comment by SteveL-MSFT** [comment]
@rjmholt You raise a fair point.  However, I think most cases the user doesn't need a complete object.  In the issue, the problem with the current behavior is that the `NestedModules` property gets cleared out when used with `Update-ModuleManifest` which is dataloss for the customer.  The fake moduleinfo is sufficient as the modulemanifest is really just a hashtable.  Although it seems I should add a test to Update-ModuleManifest.

I would suggest deferring whether we want to add capability to get the entire tree of objects until there is a customer request for it.

**Comment by daxian-dbw** [comment]
@SteveL-MSFT Thanks for fixing this regression! It looks good to me, but since the same code was both here and for `requiredModules`, would it make sense to have a private helper method to encapsulate the code for creating the fake `ModuleInfo` object?

The helper method can be a iterator method, which can be used for both here an `requiredModules`:
```
private IEnumerable<ModuleInfo> CreateFakeModuleObject(IEnumerable<ModuleSpecification> moduleSpecs)
{
    foreach (ModuleSpecification moduleSpec in moduleSpecs)
    {
        var fakeModuleInfo = new PSModuleInfo(moduleSpec.Name, Context, null);
        if (moduleSpec.Guid.HasValue)
        {
            fakeModuleInfo.SetGuid(moduleSpec.Guid.Value);
        }
        fakeModuleInfo.SetVersion(moduleSpec.RequiredVersion ?? moduleSpec.Version);
        yield fakeModuleInfo;
    }
}
```

**Comment by SteveL-MSFT** [comment]
@daxian-dbw yes, I'll make that change.  Thanks!

**Review by daxian-dbw — APPROVED** [review]
LGTM

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 2]
```diff
@@ -1338,6 +1338,25 @@ internal bool LoadModuleManifestData(ExternalScriptInfo scriptInfo, ManifestProc
             return true;
         }
 
+        /// <summary>
+        /// Helper function to generate fake PSModuleInfo objects from ModuleSpecification objects
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: Please add final dot.

File: test/powershell/engine/Module/TestModuleManifest.Tests.ps1 [hunk 3]
```diff
@@ -2,13 +2,17 @@
 # Licensed under the MIT License.
 Describe "Test-ModuleManifest tests" -tags "CI" {
 
+    BeforeEach {
+        $testModulePath = "testdrive:/module/test.psd1"
+        New-Item -ItemType Directory -Path testdrive:/module
```
Comment (iSazonov) [inline test/powershell/engine/Module/TestModuleManifest.Tests.ps1]: Please suppress output `New-Item -ItemType Directory -Path testdrive:/module > $null`.

File: test/powershell/engine/Module/UpdateModuleManifest.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+Describe "Update-ModuleManifest tests" -tags "CI" {
+
+    BeforeEach {
+        $testModulePath = "testdrive:/module/test.psd1"
+        New-Item -ItemType Directory -Path testdrive:/module
```
Comment (iSazonov) [inline test/powershell/engine/Module/UpdateModuleManifest.Tests.ps1]: The same.

File: test/powershell/engine/Module/UpdateModuleManifest.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        Remove-Item -Recurse -Force -ErrorAction SilentlyContinue testdrive:/module
+    }
+
+    It "Update should not clear out NestedModules: <variation>" -TestCases @(
+        @{ variation = "export with wildcards"; exportValue = "*" },
+        @{ variation = "export without wildcards"; exportValue = "@()"}
+    ) {
+        param($exportValue)
+
+        New-Item -ItemType File -Path testdrive:/module/foo.psm1
```
Comment (iSazonov) [inline test/powershell/engine/Module/UpdateModuleManifest.Tests.ps1]: The same.

**Review by iSazonov — APPROVED** [review]

**Pull request merged**